xla/service/gpu/kernels/gemm_kernel.cu.cc

```diff
@@ -0,0 +1,109 @@
+/* Copyright 2023 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "xla/service/gpu/kernels/gemm_kernel.h"
+
+#include "third_party/gpus/cutlass/include/cutlass/gemm/device/gemm.h"
+#include "xla/xla_data.pb.h"
+
+namespace xla::gpu::kernel {
+
+// The most basic CUTLASS f32 gemm kernel.
+using CutlassGemm =
+    cutlass::gemm::device::Gemm<float, cutlass::layout::RowMajor, float,
+                                cutlass::layout::RowMajor, float,
+                                cutlass::layout::RowMajor>;
+
+StatusOr<InProcessCutlassGemm> GetGemmKernel(PrimitiveType dtype, int32_t m,
+                                             int32_t n, int32_t k) {
```

Comment:
Thanks for starting out on the first example of wrapping CUTLASS gemms into XLA:GPU. Below are some thoughts and comments as this work progress further. 

In addition to expanding beyond single data type `PrimitiveType dtype`, which is mentioned [here](https://github.com/openxla/xla/pull/6842/files#diff-ddb7a6beda74747e37185c198c148db725571a51dc6e596379acc08310fa0c0eR56), would it make sense to separate compile-time and runtime arguments. The `GetGemmkernel` returns a CUTLASS-specific GEMM (`InProcessCutlassGemm`), CUTLASS GEMMs can operate on dynamic shape. Also, instead of single GEMM kernel would getting a list of functionally equivalent kernels for a `GemmSpec` would make more sense? 

Specifically, for `GetGemmKernel` function, I would see it as `StatusOr<GemmKernelList> GetGemmKernelList(GemmSpecification gemm_spec)`

Where `GemmSpecification` is describing all the GEMM parameters that can change for a CUTLASS GEMM at compile-time.
```
struct GemmSpecification {
   DataTypeId lhs;
   LayoutId lhs;
   DataTypeId rhs;
   LayoutId rhs;
   DataTypeId accumulation
   DataTypeId result;
   LayoutId result;
};
```

Many GEMM kernel can satisfy the above functional sepecicition (e.g. different ThreadblockShape), it will need an auto-tuner/heuristic to select top-performing GEMM kernel implementation from the `GemmKernelList`.

Links to some reference code from CUTLASS repo that could we be helpful as this work moves along:
- [GemmFunctionalKey](https://github.com/NVIDIA/cutlass/blob/1d7f2a207ec215e037099f4ba5632ccfa0249673/tools/library/include/cutlass/library/operation_table.h#L56)
- Using [GemmFunctionalKey](https://github.com/NVIDIA/cutlass/blob/1d7f2a207ec215e037099f4ba5632ccfa0249673/tools/library/src/handle.cu#L371) to get a list of GEMM kernels list 
- [GemmOperationBase](https://github.com/NVIDIA/cutlass/blob/1d7f2a207ec215e037099f4ba5632ccfa0249673/tools/library/src/gemm_operation.h#L58) class is a good reference based on my understanding of what we are trying to achieve below. GemmOperationBase builds a C++ wrapper around CUTLASS GEMM kernel held in the `typename Operator_`.
  - Converts compile-time kernel template arguments into a runtime description object that be used to query  at runtime to introspect GEMM kernel properties. 
  - Provides a common API for all CUTLASS GEMM kernels for checking if the kernel [can be implemented](https://github.com/NVIDIA/cutlass/blob/1d7f2a207ec215e037099f4ba5632ccfa0249673/tools/library/src/gemm_operation.h#L220) i.e. executed on the runtime arguments (checks alignment etc), [additional workspace requirement](https://github.com/NVIDIA/cutlass/blob/1d7f2a207ec215e037099f4ba5632ccfa0249673/tools/library/src/gemm_operation.h#L255) (if any), [initialize kernel op & arguments](https://github.com/NVIDIA/cutlass/blob/1d7f2a207ec215e037099f4ba5632ccfa0249673/tools/library/src/gemm_operation.h#L273), and [run](https://github.com/NVIDIA/cutlass/blob/1d7f2a207ec215e037099f4ba5632ccfa0249673/tools/library/src/gemm_operation.h#L295) the kernel.

Reply:
I reworked it a little bit to add a `CustomKernel` class that should be able to wrap any CUTLASS kernel (almost any, I still need to plumb workspace requirement).

And yes, it will be a list of compatible kernels and then auto tuning will pick the best one at compile time., something like:

```cpp
StatusOr<std::vector<CustomKernel>> GetGemmKernels(PrimitiveType dtype, ... mnk, ... epilogue);
```

In XLA we already have `GemmConfig` (https://github.com/openxla/xla/blob/d09d73cfe0f54f631131c133d0b667bd049a6a96/xla/service/gpu/matmul_utils.h#L90) that is very similar to your `GemmSpecification` example, I'll look into accepting it directly as it's passed around a lot, not yet sure how it's composable with epilogue fusions.

Indeed this custom kernel args packing is very similar to `GemmOperationBase`, we need to update CUTLASS in XLA, and I hope to reuse most of the `GemmOperationBase` directly, though it might be hard because for StreamExecutor integration I need to get a pointer to kernel entrypoint, and cutlass hides a lot of that.